### PR TITLE
=str Skip the iterator creation when iterable is empty.

### PR DESCRIPTION
**Pull request by He-Pin**
Motivation:
When the result iterable is empty, we can skip the iterator creation.

Result:
Better performance when the iterable is empty.

I benchmarked with 100% empty iterable

```
Jmh/run -i 3 -wi 3 -f1 -t1 .*StatefulMapConcatBenchmark.*

[info] StatefulMapConcatBenchmark.bench  thrpt    3  20244383.890 锟斤拷 293235.214  ops/s

baseline

[info] StatefulMapConcatBenchmark.bench  thrpt    3  18170728.177 锟斤拷 1067858.198  ops/s

```

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -2351,8 +2351,16 @@ private[akka] final class StatefulMapConcat[In, Out](val f: () => In => Iterable
 
     override def onPush(): Unit =
       try {
-        currentIterator = plainFun(grab(in)).iterator
-        pushPull(shouldResumeContext = false)
+        val iterable = plainFun(grab(in))
+        if ((iterable.isEmpty): @nowarn("msg=deprecated")) {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2355]: I don't think we should start using this as it was deprecated already for 2.13.0

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -2351,8 +2351,16 @@ private[akka] final class StatefulMapConcat[In, Out](val f: () => In => Iterable
 
     override def onPush(): Unit =
       try {
-        currentIterator = plainFun(grab(in)).iterator
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2354]: I don't think this optimization is worth adding. You would normally return `Nil` if you know that it's empty and that is already using a single shared empty iterator instance.

File: akka-bench-jmh/src/main/scala/akka/stream/StatefulMapConcatBenchmark.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  private implicit val system: ActorSystem = ActorSystem("StatefulMapConcatBenchmark", config)
+
+  @TearDown
+  def shutdown(): Unit = {
+    Await.result(system.terminate(), 5.seconds)
+  }
+
+  private val emptyArray = new Array[Int](0)
+  private val statefulMapConcatEmptyArray =
+    Source.repeat(0).take(OperationsPerInvocation).mapConcat(_ => emptyArray).toMat(Sink.ignore)(Keep.right)
```
Comment (patriknw) [inline akka-bench-jmh/src/main/scala/akka/stream/StatefulMapConcatBenchmark.scala:46]: more realistic test would be to use `Nil` and then I doubt that you will notice any difference

Comment (He-Pin) [reply akka-bench-jmh/src/main/scala/akka/stream/StatefulMapConcatBenchmark.scala:46]: Yes, for `Nil` will make no difference.

**Comment by He-Pin** [comment]
Let's me close this, as the Array case is rare.
